### ISSUE TITLE
Token Revocation doesn't work

### ISSUE THREAD

**Issue by mekza** [lepture/flask-oauthlib#233]
Hi,

I get this stacktrace when I try to implement token revocation

```
Traceback (most recent call last):
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask_oauthlib/provider/oauth2.py", line 549, in decorated
    return f(*args, **kwargs)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask_oauthlib/provider/oauth2.py", line 524, in decorated
    uri, headers=headers, body=body, http_method=http_method)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/oauthlib/oauth2/rfc6749/endpoints/base.py", line 64, in wrapper
    return f(endpoint, uri, *args, **kwargs)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/oauthlib/oauth2/rfc6749/endpoints/revocation.py", line 76, in create_revocation_response
    request.token_type_hint, request)
  File "/Users/martin/Dev/Bright/api/lib/python3.4/site-packages/flask_oauthlib/provider/oauth2.py", line 951, in revoke_token
    if tok and tok.client_id == request.client.client_id:
AttributeError: 'NoneType' object has no attribute 'client_id'
```

Does anyone have ever implement this?

**Comment by mekza** [comment]
Bug found in ```utils.decode_base64```

**Comment by lepture** [comment]
@mekza could you add an test case for it?

**Comment by mekza** [comment]
@lepture sure.

**Comment by ploxiln** [comment]
I also ran into this. I previously was using Flask-OAuthlib==0.7.0 for this application and just updated it to Flask-OAuthlib==0.9.2, and that's when this started. As noted by https://github.com/lepture/flask-oauthlib/pull/234#issuecomment-133856812 adding `&grant_type=password` to the request works around it.

My backtrace is pretty much the same, although I'm using python-2.7.

**Comment by ploxiln** [comment]
This change in behavior is actually not due to the Flask-OAuthlib exactly, it's due to a change of oauthlib between 0.6.3 to 1.0.3. (Downgrading Flask-OAuthlib didn't fix this, but downgrading oauthlib did fix this for me.)

Of course it's hard to say which lib is really in error, or simply exposed a bug in the other :)

**Comment by lepture** [comment]
@ploxiln could you open an issue on OAuthlib

**Comment by mekza** [comment]
:+1: Sorry I didn't find the time to work on it since this summer.

**Comment by ploxiln** [comment]
I haven't really looked into what exactly happens between Flask-OAuthlib and OAuthlib, and I'd want to do that and have a bit more evidence of a bug in OAuthlib before opening a bug there. After all, there wasn't a promise of backwards compatibility before 1.0

I'll likely do so in a few days :)

**Comment by ploxiln** [comment]
I looked into this more.

My suspicion about "revocation endpoint will now return empty string in response body" was wrong, that happens after where the exception / backtrace starts.

The change in oauthlib that causes this change in behavior is this:

oauthlib 0.7.2:
```
    def validate_revocation_request(self, request):
        """..."""
        if not request.token:
            raise InvalidRequestError(request=request,
                                      description='Missing token parameter.')

        if not self.request_validator.authenticate_client(request):
            raise InvalidClientError(request=request)
...
```

oauthlib 1.0.3:
```
    def validate_revocation_request(self, request):
        """..."""
        if not request.token:
            raise InvalidRequestError(request=request,
                                      description='Missing token parameter.')

        if self.request_validator.client_authentication_required(request):
            if not self.request_validator.authenticate_client(request):
                raise InvalidClientError(request=request)
...
```

So it's actually the fact that `client_authentication_required()` in Flask-Oauthlib is now being called, and that always returns False if grant_type is not included in the request, so then `authenticate_client()` is not called, so it doesn't set `request.client`.

I'm not sure if this is a good idea, but it looks like another possible fix is changing the line that throws the exception to `if tok and tok.client_id == request.client_id:` because it looks like (for my cases) client_id does exist in the request object (because it's in the POST body, for my application anyway).

**Comment by matt3o** [comment]
I'm still running into this issue. Has this been fixed in a newer release / has anyone filed a bug report?

**Comment by frol** [comment]
This seems to be fixed.

**Comment by w-bin** [comment]
yes, just  `pip install --upgrade flask-oauth` 
this bug seems to be fixed in release 0.9.4.  #290  : )

**Comment by mekza** [comment]
I am closing the issue then.
